networking_cisco/plugins/cisco/cfg_agent/service_helpers/routing_svc_helper_aci.py

```diff
@@ -0,0 +1,217 @@
+            LOG.warning(e)
+        except cfg_exceptions.DriverException as e:
+            with excutils.save_and_reraise_exception():
+                self.updated_routers.update([ri.router_id])
+                LOG.error(e)
+
+    def _add_rid_to_vrf_list(self, ri):
+        """Add router ID to a VRF list.
+
+        In order to properly manage VRFs in the ASR, their
+        usage has to be tracked. VRFs are provided with neutron
+        router objects in their hosting_info fields of the gateway ports.
+        This means that the VRF is only available when the gateway port
+        of the router is set. VRFs can span routers, and even OpenStack
+        tenants, so lists of routers that belong to the same VRF are
+        kept in a dictionary, with the VRF name as the key.
+        """
+        if ri.ex_gw_port or ri.router.get('gw_port'):
+            driver = self.driver_manager.get_driver(ri.id)
+            vrf_name = driver._get_vrf_name(ri)
+            if not vrf_name:
+                return
+            if not self._router_ids_by_vrf.get(vrf_name):
+                LOG.debug("++ CREATING VRF %s" % vrf_name)
+                driver._do_create_vrf(vrf_name)
+            self._router_ids_by_vrf.setdefault(vrf_name, set()).add(
+                ri.router['id'])
+
+    def _remove_rid_from_vrf_list(self, ri):
+        """Remove router ID from a VRF list.
+
+        In order to properly manage VRFs in the ASR, their
+        usage has to be tracked. VRFs are provided with neutron
+        router objects in their hosting_info fields of the gateway ports.
+        This means that the VRF is only available when the gateway port
+        of the router is set. VRFs can span routers, and even OpenStack
+        tenants, so lists of routers that belong to the same VRF are
+        kept in a dictionary, with the VRF name as the key.
+        """
+        if ri.ex_gw_port or ri.router.get('gw_port'):
+            driver = self.driver_manager.get_driver(ri.id)
+            vrf_name = driver._get_vrf_name(ri)
+            if self._router_ids_by_vrf.get(vrf_name) and (
+                    ri.router['id'] in self._router_ids_by_vrf[vrf_name]):
+                self._router_ids_by_vrf[vrf_name].remove(ri.router['id'])
+            # If htis is the last router in a VRF, then we can safely
+            # delete the VRF from the router config (handled by the driver)
+            if not self._router_ids_by_vrf.get(vrf_name):
+                LOG.debug("++ REMOVING VRF %s" % vrf_name)
+                driver._remove_vrf(ri)
+                if self._router_ids_by_vrf.get(vrf_name) is not None:
+                    del self._router_ids_by_vrf[vrf_name]
+
+    def _router_removed(self, router_id, deconfigure=True):
+        """Specialize _router_removed.
+
+        This adds a check to see if we missed handling the VRF
+        (could happen if the transition of clearing the gateway
+        was somehow missed).
```

Comment:
why this transition will be missed? It sounds like a bug to me if that happens.....